internal/serving_helper.go

```diff
@@ -0,0 +1,152 @@
+	}
+}
+
+func GetCreateKsvcFunc(ksvcClient servingv1client.ServingV1Interface, minScale, maxScale int, svcPrefix string, timeout time.Duration) func(ns string, index int) (string, string) {
+	return func(ns string, index int) (string, string) {
+		service := servingv1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("%s-%d", svcPrefix, index),
+				Namespace: ns,
+			},
+		}
+
+		service.Spec.Template = servingv1.RevisionTemplateSpec{
+			Spec: servingv1.RevisionSpec{},
+			ObjectMeta: metav1.ObjectMeta{
+				Annotations: map[string]string{
+					"autoscaling.knative.dev/minScale": strconv.Itoa(minScale),
+					"autoscaling.knative.dev/maxScale": strconv.Itoa(maxScale),
+				},
+			},
+		}
+		service.Spec.Template.Spec.Containers = []corev1.Container{
+			{
+				Image: ServiceImage,
+				Ports: []corev1.ContainerPort{
+					{
+						ContainerPort: 8080,
+					},
+				},
+			},
+		}
+		fmt.Printf("Creating Knative Service %s in namespace %s\n", service.GetName(), service.GetNamespace())
+		_, err := ksvcClient.Services(ns).Create(context.TODO(), &service, metav1.CreateOptions{})
+		if err != nil {
+			fmt.Printf("failed to create Knative Service %s in namespace %s : %s\n", service.GetName(), service.GetNamespace(), err)
+		}
+		return service.GetNamespace(), service.GetName()
+	}
+}
+
+func GetListServicesFunc() GetServicesFunc {
+	return func(servingClient servingv1client.ServingV1Interface, nsNameList []string, svcPrefix string) []ServicesToScale {
+		objs := []ServicesToScale{}
+		for _, ns := range nsNameList {
+			svcList, err := servingClient.Services(ns).List(context.TODO(), metav1.ListOptions{})
+			if err == nil {
+				for _, s := range svcList.Items {
+					svc := s
+					if strings.HasPrefix(s.Name, svcPrefix) {
+						objs = append(objs, ServicesToScale{Namespace: ns, Service: &svc})
+					}
+				}
+			}
+		}
+		return objs
+	}
+}
+
+func GetRevisionReadyCondition() apis.ConditionType {
+	return servingv1.RevisionConditionReady
+}
+
+func GetServiceConditionConfigurationsReady() apis.ConditionType {
+	return servingv1.ServiceConditionConfigurationsReady
+}
+
+func GetServiceConditionRoutesReadyType() apis.ConditionType {
+	return servingv1.ServiceConditionRoutesReady
+}
+
+func GetActivatorEndpointsPopulatedCondition() apis.ConditionType {
+	return networkingv1.ActivatorEndpointsPopulated
+}
+func GetServerlessServiceConditionEndspointsPopulated() apis.ConditionType {
+	return networkingv1.ServerlessServiceConditionEndspointsPopulated
+}
+
+func GetServerlessServiceConditionReady() apis.ConditionType {
+	return networkingv1.ServerlessServiceConditionReady
+}
+
+func GetIngressConditionNetworkConfigured() apis.ConditionType {
+	return networkingv1.IngressConditionNetworkConfigured
+}
+
+func GetIngressConditionLoadBalancerReady() apis.ConditionType {
+	return networkingv1.IngressConditionLoadBalancerReady
+}
+
+func GetPodAutoscalerConditionActive() apis.ConditionType {
```

Comment:
When I tried to use kperf in knative, I get a circular dependency since kperf uses serving and now serving would use knative, so one solution I though of was to move all the serving dependency to an internal package here

Another option is to just use kperf an a binary in serving

Reply:
I see, yeah, I'd be interested in other opinions/reasons to have it be a library - I could totally be missing an angle here - but certainly in my head I was imagining kperf would be a binary we'd run via a prow job/action rather than a library. If we do want to make it a library I wonder if there might be an easier way of breaking the circular dep, for example calling it from a subpackage of test without dependencies on serving (eg knative.dev/serving/test/performance).

Reply:
I wanted to use as a library to make writing the actual tests easier, https://github.com/knative/serving/pull/12272 before realizing that circular dependency

Reply:
but now open to using it as a binary, but as far as I understand would still be calling that binary from the go code in the test

Reply:
FWIW I'm pretty convinced we can find a better way of doing this rather than having these wrapper functions here, even if we do want to use this as a library. Even if serving/test/performance relies on kperf I don't see why that's a problem unless kperf relies on the serving/test/performance package, and I don't see why it has to do that (separately we're going to have a painful dependency in the release process when we try to pin versions since the repos rely on each other, but the internal package doesn't help with that at all, unless I'm missing something?). Could you share a link to a broken build with the non-internal version so I can have a look?

Reply:
yeah, I don't think going down the route of the internal package would have helped anyways. I have switched the tests in serving#12272 to just use a kperf binary as you suggested, so I don't think we need this anymore. 
It was more of a trail to see if we contain the serving  imports in one package, it would not be a circular dependency in serving, but it turns out the common packages also such as `cache` and `pkg` could be an issue if they have a different version.